docs/design_proposals/filesystem-resolution-proposal-01.md

```diff
@@ -0,0 +1,110 @@
+* Reviewers:
+* Date: 2020-02-12
+* Status: [Reviewed/Cancelled/Under implementation/Complete]
+
+Here is a brief explanation of the Statuses
+
+1. Reviewed: The proposal PR has been accepted, merged and ready for
+   implementation.
+2. Under implementation: An accepted proposal is being implemented by actual work.
+   Note: The design might change in this phase based on issues during
+   implementation.
+3. Cancelled: During or before implementation the proposal was cancelled.
+   It could be due to:
+   * other features added which made the current design proposal obsolete.
+   * No longer a priority.
+4. Complete: This feature/change is implemented.
+
+## Background
+
+Kaniko builds Docker image layers as overlay filesystem layers; specifically it
+creates a tar file which contains the entire content of a given layer in the
+overlay filesystem. Each overlay layer corresponds to one image layer.
+
+Overlay filesystems should only contain the objects changed in each layer;
+meaning that if only one file changes between some layer A and some B, layer B
+would only contain a single file (the one that changed).
+
+To accomplish this, Kaniko walks the entire filesystem to discover every object.
+Some of these objects may actually be a symlink to another object in the
+filesystem; in these cases we must consider both the link and the target object.
+
+Kaniko also maintains a set of whitelisted (aka ignored) filepaths. Any object
+which matches one of these filepaths should be ignored by kaniko.
+
+This results in a 3 dimensional search space
+
+* changed relative to previous layer
+* symlink
+* whitelisted
+
+Kaniko must also track which objects are referred to by multiple stages; this
+functionality is out of scope for this proposal.
+
+This search space is currently managed in an inconsistent and somewhat ad-hoc
+way; code that manages the various search dimensions is spread out and
+duplicated. There are also a number of poorly handled edge cases which continue
+to cause bugs.
+
+The search space dimensions cannot be reduced or substituted.
+
+Currently there are a number of bugs around symlinks incorrectly resolved,
+whitelists not respected, and unchanged files added to layers.
+
+## Design
+
+Resolution of the filesystem and the objects it contains should be handled
+consistently and with a single API.
+
+* Callers of this API should not be concerned with the type of object at a given filepath (e.g. symlink or not).
+* Callers of this API should not be concerned with whether a given path is whitelisted.
+* This API should return a set of filepaths which should be added to the layer
+  without further processing.
+
+The API should take a limited set of arguments
+* The root path in the filesystem
+* The whitelist
+
+The API should return only two arguments
+* A set of filepaths which should be added to the layer without further
+  processing
+* error
+
+### Open Issues/Questions
+
+\<Ignore symlinks targeting whitelisted paths?\>
```

Comment:
I guess this is related to copy question as well https://github.com/GoogleContainerTools/kaniko/pull/1048/files#r378600747

If a symlink targets a whitelisted path, should we ignore that symlink?

If the destination of a copy command is a whitelisted path, should we ignore that command?

Reply:
Lets leave looking into `Copy` and `Add` from this proposal and only focus on API for snapshotting entire file snapshot. 

For the snapshot API, we *should not* ignore whitelisted paths.
e.g. An image 'gcr.io/google/toolprovider" provides binary `/usr/bin/tool_binary` which is symlinked to whitelisted path `var/run/tool_binary`, the right thing to do is create a file `/usr/bin/tool_binary` with content of `/var/run/tool_binary`.
We should ignore `/var/run/tool_binary` from the snapshot.
Does that make sense to you?

Reply:
It sounds like you are proposing a new behavior, which is to not add symlinks to the layer tar, but instead to create regular files with the content of the symlink target (basically what you mentioned the other day). Is that correct?

Reply:
yes. Docker does the same. It does not create the symlink.

Reply:
that seems reasonable, but also perhaps out of scope for this proposal (specifically how a file is added to the tar).

As far as this proposal specifically, we need to consider whether to ignore paths which are symlinks that target whitelisted objects.

It sounds like you are suggesting, no we should not ignore them.